More robust treatment of the Discharge status

herbelin opened the pull request:
**Kind:** cleanup / bug fix

Has overlays to be merged synchronously:
- LPCIC/coq-elpi#62
- mattam82/Coq-Equations#210
- Mtac2/Mtac2#210
- mit-plv/fiat#27
- coq-community/paramcoq#32

<!-- If this is a bug fix, make sure the bug was reported beforehand. -->
Fixes #8722

Update: It was intended to also fix #8725 but the latter has been fixed differently.

- [X] Added / updated test-suite
- [ ] Entry to be added in CHANGES.md? I have no idea whether one can fall on these side corner examples by chance, and rely on them. [Update: personally, I would not add a comment to the coq-level change log but I can if at least someone insist.]

See [this](https://github.com/coq/coq/pull/8717#issuecomment-429317576) comment for some questions on the design issues. The names of constructors are a bit long but I could not find one both short and non too ambiguous. [Update: I found a simplification of the names of constructors. See [below](https://github.com/coq/coq/pull/8726#issuecomment-497938838).]

See #8722 and #8725 to evaluate whether the reported behaviors have indeed to be fixed or not.

I changed the message `id is declared as a local definition` into the much longer `Interpreting this declaration as if a global declaration prefixed by "Local" i.e. as a global declaration which shall not be available without qualification when imported.`. The point is that the message is now given at the step going from parser to command interpreter. I drop the name of `id` but I can reintroduce it if there is a demand. Suggestions to simplify the message welcome. [Update: the message seems then ok.]

@ejgallego: Where do you suggest I put the locality type? Maybe in `declare.ml`? [Update: I kept them in `decl_kinds.ml` which seems ok.]

ejgallego commented:
@herbelin I've moved the locality info to declaredef , let me finish that patch, then I look at this one, and I rebase on top.

Won't happen today tho.

ppedrot commented:
@ejgallego Are you assigning this to you? Or should I?

ppedrot commented:
@ejgallego Are you assigning this to you? Or should I?

ppedrot commented:
Are the CI failures related? I don't know whether there has been some discrepancy between the devs and the master branch...

herbelin commented:
Failure of Mtac2 overlay seems similar to what @ppedrot reported [here](https://github.com/coq/coq/pull/7925#issuecomment-434262470):
```
Fatal error: exception Sys_error("$(COQMF_COQLIB)/user-contrib/Unicoq: 
No such file or directory")
```

ppedrot commented:
@herbelin I found out the problem, it is due to the Mtac2 branch for the CI. Instead of master, the PR needs to be based on master-sync.

herbelin commented:
@ppedrot: thanks.

So, this PR has 5 overlays. Obviously, with `Declare.declare_definition`, we are touching at a key function of the API for plugin.

Maybe worth to simplify further the API? I'm a bit worry by the complexity of a so-called "simple declare" in `tuto1`...

SkySkimmer commented:
>So, this PR has 5 overlays. Obviously, with Declare.declare_definition, we are touching at a key function of the API for plugin.

@herbelin did you intend to comment in #8718? This PR doesn't touch declare_definition's API AFAICT.

herbelin commented:
> This PR doesn't touch declare_definition's API AFAICT.

It does: one of argument has a different type.

herbelin commented:
This should be OK. Rebased and double-checking CI.

(In passing, if someone has an idea for a ligher name for the constructors of `type import_status = ImportAllowUnqualified | ImportNeedQualified`... otherwise, let's stay as it is.)

maximedenes commented:
> Fixes #8722, fixes #8725

The PR does not seem to contain test files in `bugs`. You can have a look at #9328, which has one.

vbgl commented:
Hugo, could you please rebase to fix the conflicts and add the missing test case? Or update the milestone if this change does not belong to version 8.10.

Thanks.

vbgl commented:
Rebased. Please double-check, in particular the only part where I had to be creative: https://github.com/coq/coq/pull/8726/commits/e1e1732574366ed0ab2509db1142bda7566e30b8#diff-e740c45ba76a0ca63da5a9b470a55ac3R106

The test case is still missing.

Overlays are probably out-of-date.

ejgallego commented:
Looks fine to me, even if that's a pity that the proof rework tree will have to undo some of the moves here due to the reification of the control flow.

Regarding overlays I dunno, I could always include this PR in the proof rework tree and take care of them this way.

vbgl commented:
Added a test-case.

vbgl commented:
Fixed the plugin-tutorial.

herbelin commented:
@vgbl: thanks for the rebasing and the test suite. About your question in `abstract.ml`, it should be `ImportAllowUnqualified` (I pushed the change).

Actually, I'm not proud of the terminology:
- `ImportAllowUnqualified` is supposed to mean: available at import without qualification (was the former meaning of `Global` which I found misguiding because not suggesting a difference of status after `Import`)
- `ImportNeedQualified` is supposed to mean: needs to be qualified even after `Import` (was the former meaning of `Local` which itself was somehow source of confusion with discharged)

Maybe `AvailableAfterImportWithoutQualification` and `NeedsQualificationEvenAfterImport` would be clearer? (But it is so long...).

Overlays: I updated the overlays for Equations, Mtac2, Elpi. I added an overlay for paramcoq (but w/o making a PR).

ejgallego commented:
> * `ImportNeedQualified` is supposed to mean: needs to be qualified even after `Import` (was the former meaning of `Local` which itself was somehow source of confusion with discharged)

Thanks for the explanations @herbelin , what are such cases, for Local inside sections?

SkySkimmer commented:
>Thanks for the explanations @herbelin , what are such cases, for Local inside sections?

In the v file syntax `Local Definition` is ImportNeedQualified
Recall that the meaning of user syntax Local/Global is highly dependent on the command it's attached to.

ejgallego commented:
> In the v file syntax `Local Definition` is ImportNeedQualified
> Recall that the meaning of user syntax Local/Global is highly dependent on the command it's attached to.

I see, thanks @SkySkimmer , that confuses me a bit, and I would think that `Local Definition` would mean "not import at all", à la Haskell.

ejgallego commented:
> I see, thanks @SkySkimmer , that confuses me a bit, and I would think that `Local Definition` would mean "not import at all", _à la_ Haskell.

Of course you need to close the dependency graph of definitions / proofs in our case; and that may be annoying.

ejgallego commented:
Is there some model of module systems with "hiding", I guess definition depending on private constants could be closed with an existential, in fact this is already done on purpose by some libraries to really hide stuff.

herbelin commented:
This PR was adding a new warning to be consistent with a missing case of the "definition-not-visible" warning (a `Let` made with tactics). But this warning was removed in #10188 so, I propose to remove the new warning (called `warn_statement_not_visible` in the PR) also to remain consistent with the new absence of warning. @skyskimmer, I guess this looks ok to you.

SkySkimmer commented:
OK with me

herbelin commented:
> OK with me

OK, I removed it.

I also made a last change. I was not really satisfied with the replacement of `Global` by the long and a bit headache-name `Persistent ImportAllowUnqualified`. I replaced it with `Global ImportDefaultBehavior` so that users who wants an easy migration path canonically replace `Global` by `Global ImportDefaultBehavior`. Of course, it depends on the intended meaning of the vernac command `Local Definition` and of how we want or can implement it, since it is not clear how to implement it in a stronger way than by only briddling the ability to qualify the name.

herbelin commented:
This seems technically ready: pending comments addressed, CI green, PRs for overlays submitted.

herbelin commented:
The possible criticism I see to this PR is the number of plugin overlays it needs. But based on the basis that API are still not stable, I'd assume that it is acceptable.

@ppedrot, you assigned apparently.

herbelin commented:
Note: once more time rebased.

Zimmi48 commented:
> I changed the message `id is declared as a local definition` into the much longer `Interpreting this declaration as if a global declaration prefixed by "Local" i.e. as a global declaration which shall not be available without qualification when imported.`. The point is that the message is now given at the step going from parser to command interpreter. I drop the name of `id` but I can reintroduce it if there is a demand. Suggestions to simplify the message welcome. [Update: the message seems then ok.]

I had not seen this but the new message seems overly complex. I will open a PR to simplify it. Furthermore, the documentation should have been updated. I will do this as well.

JasonGross commented:
I think the identifier should appear in the warning, otherwise it is nearly useless on the command line
